pkg/log/log.go

```diff
@@ -0,0 +1,235 @@
+	//
+	// The logger instantiated here can be changed to any logger
+	// implementing the logr.Logger interface. This logger will
+	// be propagated through the whole operator, generating
+	// uniform and structured logs.
+	logf.SetLogger(zap.Logger())
+
+	once.Do(func() {
+		logger = newLogger(name, out, development)
+	})
+}
+
+func newLogger(name string, out io.Writer, development bool) *Logger {
+	return &Logger{
+		logr: logf.ZapLoggerTo(out, development).WithName(name),
+	}
+}
+
+// Info logs a non-error message.
+func Info(ctx *gin.Context, msg string) {
+	logger.Info(ctx, msg)
+}
+
+// Infof logs a non-error formatted message.
+func Infof(ctx *gin.Context, msg string, args ...string) {
+	logger.Infof(ctx, msg, args...)
+}
+
+// Error logs the error with the given message.
+func Error(ctx *gin.Context, err error, msg string) {
+	logger.Error(ctx, err, msg)
+}
+
+// Errorf logs the error with the given formatted message.
+func Errorf(ctx *gin.Context, err error, msg string, args ...string) {
+	logger.Errorf(ctx, err, msg, args...)
+}
+
+// WithValues creates a new logger with additional key-value pairs in the context
+func WithValues(keysAndValues map[string]interface{}) *Logger {
+	return logger.WithValues(keysAndValues)
+}
+
+// Info logs are used for non-error messages. It will log a message with
+// the given key/value pairs as context.
+func (l *Logger) Info(ctx *gin.Context, msg string) {
+	logger.Infof(ctx, msg)
+}
+
+// Infof logs are used for non-error messages. It will log a message with
+// the given key/value pairs as context.
+func (l *Logger) Infof(ctx *gin.Context, msg string, args ...string) {
+	ctxInfo := addContextInfo(ctx)
+	arguments := make([]interface{}, len(args))
+	for i, arg := range args {
+		arguments[i] = arg
+	}
+	if len(arguments) > 0 {
+		l.logr.Info(fmt.Sprintf(msg, arguments...), ctxInfo...)
+	} else {
+		l.logr.Info(msg, ctxInfo...)
+	}
+}
+
+// Error logs are used for logging errors. It will log the error with the given
+// message and key/value pairs as context.
+func (l *Logger) Error(ctx *gin.Context, err error, msg string) {
+	l.Errorf(ctx, err, msg)
+}
+
+// Errorf logs are used for logging errors. It will log the error with the given
+// message and key/value pairs as context.
+func (l *Logger) Errorf(ctx *gin.Context, err error, msg string, args ...string) {
+	ctxInfo := addContextInfo(ctx)
+	arguments := make([]interface{}, len(args))
+	for i, arg := range args {
+		arguments[i] = arg
+	}
+
+	if len(arguments) > 0 {
+		l.logr.Error(err, fmt.Sprintf(msg, arguments...), ctxInfo...)
+	} else {
+		l.logr.Error(err, msg, ctxInfo...)
+	}
+}
+
+// WithValues appends tags to the logger.
+func (l *Logger) WithValues(keysAndValues map[string]interface{}) *Logger {
+	if len(keysAndValues) > 0 {
+		// ZapLoggerTo, name and WithValues all return new logger instances.
+		// The logger must be set again with the values stored in the Logger struct.
+		logger.logr = logger.logr.WithValues(slice(keysAndValues)...)
```

Comment:
It's the same error. We can't modify the existing logger! Create a new one and return it. Don't modify `logger`. See https://github.com/tinakurian/registration-service/pull/5/files#diff-1135f265d25e96e557b8114567653176R119
Otherwise there will be race condition because multiple go routines will be constantly adding their values to the singleton logger.

Reply:
Done!